From 9e612272b0f2ae0f3fb0e2163fdcc918bc6731a4 Mon Sep 17 00:00:00 2001
From: Niall Pemberton <niallp@apache.org>
Date: Wed, 29 Sep 2010 18:51:30 +0000
Subject: [PATCH] IO-248 FilenameUtils - getFullPathNoEndSeparator returns
 empty while path is a one level directory - thanks to hippo spark for the
 patch

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/io/trunk@1002806 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/io/FilenameUtils.java   |  3 +++
 .../commons/io/FilenameUtilsTestCase.java      | 18 ++++++++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/src/java/org/apache/commons/io/FilenameUtils.java b/src/java/org/apache/commons/io/FilenameUtils.java
index 01fcadbf7e6..a61c60058be 100644
--- a/src/java/org/apache/commons/io/FilenameUtils.java
+++ b/src/java/org/apache/commons/io/FilenameUtils.java
@@ -887,6 +887,9 @@ private static String doGetFullPath(String filename, boolean includeSeparator) {
             return filename.substring(0, prefix);
         }
         int end = index + (includeSeparator ?  1 : 0);
+        if (end == 0) {
+            end++;
+        }
         return filename.substring(0, end);
     }
 
diff --git a/src/test/org/apache/commons/io/FilenameUtilsTestCase.java b/src/test/org/apache/commons/io/FilenameUtilsTestCase.java
index ba08eb06953..02232780745 100644
--- a/src/test/org/apache/commons/io/FilenameUtilsTestCase.java
+++ b/src/test/org/apache/commons/io/FilenameUtilsTestCase.java
@@ -704,6 +704,24 @@ public void testGetFullPathNoEndSeparator() {
         assertEquals("~user/a/b", FilenameUtils.getFullPathNoEndSeparator("~user/a/b/c.txt"));
     }
 
+    /**
+     * Test for https://issues.apache.org/jira/browse/IO-248
+     */
+    public void testGetFullPathNoEndSeparator_IO_248() {
+
+        // Test single separator
+        assertEquals("/",  FilenameUtils.getFullPathNoEndSeparator("/"));
+        assertEquals("\\", FilenameUtils.getFullPathNoEndSeparator("\\"));
+
+        // Test one level directory
+        assertEquals("/",  FilenameUtils.getFullPathNoEndSeparator("/abc"));
+        assertEquals("\\", FilenameUtils.getFullPathNoEndSeparator("\\abc"));
+
+        // Test one level directory
+        assertEquals("/abc",  FilenameUtils.getFullPathNoEndSeparator("/abc/xyz"));
+        assertEquals("\\abc", FilenameUtils.getFullPathNoEndSeparator("\\abc\\xyz"));
+    }
+
     public void testGetName() {
         assertEquals(null, FilenameUtils.getName(null));
         assertEquals("noseperator.inthispath", FilenameUtils.getName("noseperator.inthispath"));
